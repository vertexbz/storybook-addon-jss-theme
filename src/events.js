// @flow
export const SelectTheme = 'selectTheme';
export const SetThemes = 'setThemes';
