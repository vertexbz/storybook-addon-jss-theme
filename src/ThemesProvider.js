// @flow
import * as React from 'react';
import addons from '@storybook/addons';
import { ThemeProvider as JssThemeProvider } from 'react-jss';
import * as event from './events';
import type { ThemeType, ThemeProviderComponent } from './types';

type ThemesProviderProps = {
    ThemeProvider: ThemeProviderComponent,
    themes: Array<ThemeType>,
    children: React.Element<*>
};


type ThemesProviderState = {
    theme: ?ThemeType
};

export default
class ThemesProvider extends React.Component<ThemesProviderProps, ThemesProviderState> {
    static defaultProps = {
        ThemeProvider: JssThemeProvider
    };

    state = {
        theme: this.props.themes[0]
    };

    constructor(...args: *) {
        super(...args);

        // $FlowIgnore
        this.setTheme = this.setTheme.bind(this);
    }

    setTheme(theme: ThemeType) {
        this.setState({ theme });
    }

    componentDidMount() {
        const channel = addons.getChannel();
        channel.on(event.SelectTheme, this.setTheme);
        channel.emit(event.SetThemes, this.props.themes);
    }

    componentWillUnmount() {
        const channel = addons.getChannel();
        channel.removeListener(event.SelectTheme, this.setTheme);
    }

    render(): ?React.Element<ThemeProviderComponent> {
        if (!this.state.theme) {
            return null;
        }

        const ThemeProvider = this.props.ThemeProvider;

        return (
            <ThemeProvider theme={this.state.theme.variables}>
                {this.props.children}
            </ThemeProvider>
        );
    }
}
