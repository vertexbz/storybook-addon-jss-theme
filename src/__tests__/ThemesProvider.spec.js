import { mount } from 'enzyme';
import * as React from 'react';
import ThemesProvider from '../ThemesProvider';
import * as event from '../events';

jest.mock('@storybook/addons');
const addonsMock = require('@storybook/addons').default;

describe('ThemesProvider', () => {
    // eslint-disable-next-line max-statements
    it('should work properly', () => {
        const themes = [
            {
                name: 'theme 1',
                variables: {
                    var: '1'
                }
            },
            {
                name: 'theme 2',
                variables: {
                    var: '2'
                }
            }
        ];

        const TestThemesProvider = ({ children }) => children;
        const TestStory = () => <div />;

        const wrapper = mount(
            <ThemesProvider themes={themes} ThemeProvider={TestThemesProvider}>
                <TestStory />
            </ThemesProvider>
        );

        expect(wrapper.render()).toMatchSnapshot();
        expect(addonsMock.__channel.on).toBeCalledTimes(1);

        const registerSelectThemeCall = addonsMock.__channel.on.mock.calls[0];
        const registerSelectThemeCallback = registerSelectThemeCall[1];
        expect(registerSelectThemeCall[0]).toBe(event.SelectTheme);

        expect(addonsMock.__channel.emit).toBeCalledTimes(1);
        expect(addonsMock.__channel.emit).toBeCalledWith(event.SetThemes, themes);

        registerSelectThemeCallback(themes[0]);
        wrapper.update();
        expect(wrapper.render()).toMatchSnapshot();

        expect(wrapper.find(TestThemesProvider).exists()).toBeTruthy();
        expect(wrapper.find(TestThemesProvider).prop('theme')).toBe(themes[0].variables);

        expect(wrapper.find(TestThemesProvider).find(TestStory).exists()).toBeTruthy();



        registerSelectThemeCallback(themes[1]);
        wrapper.update();
        expect(wrapper.render()).toMatchSnapshot();

        expect(wrapper.find(TestThemesProvider).exists()).toBeTruthy();
        expect(wrapper.find(TestThemesProvider).prop('theme')).toBe(themes[1].variables);

        expect(wrapper.find(TestThemesProvider).find(TestStory).exists()).toBeTruthy();


        wrapper.unmount();
        expect(addonsMock.__channel.removeListener).toBeCalledTimes(1);
    });
});
