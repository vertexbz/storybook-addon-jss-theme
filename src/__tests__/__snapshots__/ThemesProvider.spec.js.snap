// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemesProvider should work properly 1`] = `
"<ThemesProvider themes={{...}} ThemeProvider={[Function: TestThemesProvider]}>
  <TestThemesProvider theme={{...}}>
    <div var=\\"1\\">
      <TestStory>
        <div />
      </TestStory>
    </div>
  </TestThemesProvider>
</ThemesProvider>"
`;

exports[`ThemesProvider should work properly 2`] = `
"<ThemesProvider themes={{...}} ThemeProvider={[Function: TestThemesProvider]}>
  <TestThemesProvider theme={{...}}>
    <div var=\\"1\\">
      <TestStory>
        <div />
      </TestStory>
    </div>
  </TestThemesProvider>
</ThemesProvider>"
`;

exports[`ThemesProvider should work properly 3`] = `
"<ThemesProvider themes={{...}} ThemeProvider={[Function: TestThemesProvider]}>
  <TestThemesProvider theme={{...}}>
    <div var=\\"2\\">
      <TestStory>
        <div />
      </TestStory>
    </div>
  </TestThemesProvider>
</ThemesProvider>"
`;
